examples/advanced/ocus_explanations.py

```diff
@@ -0,0 +1,302 @@
+    Celine = ~samurai
+    Anton = mayo | ~curry | ~andalouse
+    Danny = ~mayo | ketchup | andalouse | samurai
+    Luc = ~mayo | samurai
+
+    allwishes   = [Nora, Leander, Benjamin, Behrouz, Guy, Daan, Celine, Anton, Danny, Luc]
+
+    wish_weights = [40, 40, 60, 60, 60, 60, 20, 60, 80, 40]
+
+    explanation_sequence = explain_ocus(allwishes, wish_weights, hard=[], verbose=verbose)
+
+    if verbose > 0:
+        print("\nExplanation Sequence:")
+        print("-" * len("Explanation Sequence:"))
+        for id, explanation in enumerate(explanation_sequence):
+            print(f"\n{id}. Using Constraints:\t", explanation["constraints"])
+            print("\t     Derived:\t", explanation["derived"])
+            print("\t        Cost:\t", explanation["cost"])
+
+    return explanation_sequence
+
+def explain_ocus(soft, soft_weights=None,  hard=[], solver="ortools", verbose=False):
+    '''
+        A SAT-based explanation sequence generating algorithm using assumption
+        variables and weighted UNSAT core (Optimal Constrained Unsatisfiable Subset)
+        extraction. [1, 2]
+
+        [1] Bogaerts, B., Gamba, E., Claes, J., & Guns, T. (2020). Step-wise explanations
+        of constraint satisfaction problems. In ECAI 2020-24th European Conference on
+        Artificial Intelligence, 29 August-8 September 2020, Santiago de Compostela,
+        Spain, August 29-September 8, 2020-Including 10th Conference on Prestigious
+        Applications of Artificial Intelligence (PAIS 2020) (Vol. 325, pp. 640-647).
+        IOS Press; https://doi. org/10.3233/FAIA200149.
+
+        [2] Bogaerts, B., Gamba, E., & Guns, T. (2021). A framework for step-wise explaining
+        how to solve constraint satisfaction problems. Artificial Intelligence, 300, 103550.
+    '''
+    reified_soft = []
+    ind = BoolVar(shape=len(soft), name="ind")
+
+    ## adding assumption variables to the soft constraints
+    for i,bv in enumerate(ind):
+        reified_soft += [bv.implies(soft[i])]
+
+    # to map indicator variable back to soft_constraints
+    indmap = dict((v,i) for (i,v) in enumerate(ind))
+
+    # adding the assumption variables
+    I = set(bv for bv in ind)
+
+    Iend = optimal_propagate(hard + reified_soft, I, solver, verbose)
+
+    # Cost function returns cost of soft weight if a constraint is used
+    # otherwhise returns 1 (i.e. using a literal)
+    cost = cost_func(indmap, soft_weights)
+
+    if verbose > 0:
+        print("\nMAXIMAL CONSEQUENCE\n\t", Iend)
+        print("\nREMAINING TO EXPLAIN\n\t", Iend-I)
+
+    explanation_sequence = []
+
+    while(I != Iend):
+        # explain 1 step using ocus
+        ocus_expl = explain_one_step_ocus(hard+reified_soft, cost, Iend, I, solver, verbose)
+
+        # project on known facts and constraints
+        soft_used = I & ocus_expl
+
+        # propagate and find information hold in all models
+        derived = optimal_propagate(hard + reified_soft, soft_used, solver, verbose) - I
```

Comment:
for the sake of this example and cleanliness of your code, I think you can skip this; go straight to the explanation. We might generate multiple steps for the same set of constraints, but no harm is done, it is just potentially a bit less efficient.